### PR TITLE
INT-4208: Add WebFlux based HttpMessageHandler

### PR DESCRIPTION
**Pull request by artembilan**
JIRA: https://jira.spring.io/browse/INT-4208

Since `AsyncRestTemplate` is deprecated in Spring 5.0, it doesn't make
sense to promote that feature via our new `AsyncHttpRequestExecutingMessageHandler`
component

* Rework (and rename) `AsyncHttpRequestExecutingMessageHandler` to
`ReactiveHttpRequestExecutingMessageHandler` and make it based on the
WebFlux `WebClient`
* Fix Java DSL (`ReactiveHttpMessageHandlerSpec`) and all tests according a new
logic in the `ReactiveHttpRequestExecutingMessageHandler`
* Fix XML namespace support to use new `ReactiveHttpRequestExecutingMessageHandler`
and don't expose unused options like `converters` and `request-factory`
* Fix `What's New` and `http.adoc`
* To remain with `async` mode for the `ReactiveHttpRequestExecutingMessageHandler`
behavior support fix `AbstractMessageProducingHandler` to adapt reply `Mono`
to the `SettableListenableFuture`
* Introduce new `reactive` behavior for the `AbstractMessageProducingHandler`
when `outputChannel` is `ReactiveSubscribableChannel` and perform
`.subscribeTo(Publisher)` for `Publisher` reply
* Upgrade to Spring AMQP `2.0 M3`
* Downgrade to Spring Security `4.2.2`

### CODE REVIEW

**Comment by artembilan** [comment]
Ready for review

File: spring-integration-http/src/main/java/org/springframework/integration/http/outbound/ReactiveHttpRequestExecutingMessageHandler.java [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+	 */
+	public ReactiveHttpRequestExecutingMessageHandler(Expression uriExpression, WebClient webClient) {
+		super(uriExpression);
+		this.webClient = (webClient == null ? WebClient.create() : webClient);
+		this.setAsync(true);
+	}
+
+	@Override
+	public String getComponentType() {
+		return (this.getExpectReply() ? "http:outbound-reactive-gateway" : "http:outbound-reactive-channel-adapter");
```
Comment (garyrussell) [inline spring-integration-http/src/main/java/org/springframework/integration/http/outbound/ReactiveHttpRequestExecutingMessageHandler.java]: `s/this//`

File: spring-integration-http/src/main/java/org/springframework/integration/http/outbound/ReactiveHttpRequestExecutingMessageHandler.java [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+				uriSpec = this.webClient.patch();
+				break;
+			case DELETE:
+				uriSpec = this.webClient.delete();
+				break;
+			case OPTIONS:
+				uriSpec = this.webClient.options();
+				break;
+			case TRACE:
+				throw new UnsupportedOperationException("WebClient doesn't support TRACE HTTP method");
```
Comment (garyrussell) [inline spring-integration-http/src/main/java/org/springframework/integration/http/outbound/ReactiveHttpRequestExecutingMessageHandler.java]: the TRACE

File: spring-integration-http/src/main/java/org/springframework/integration/http/outbound/ReactiveHttpRequestExecutingMessageHandler.java (hunk 1 above)
Comment (garyrussell) [inline spring-integration-http/src/main/java/org/springframework/integration/http/outbound/ReactiveHttpRequestExecutingMessageHandler.java]: `s/get/is/` - looks like we missed that on the previous async review.

File: spring-integration-http/src/main/java/org/springframework/integration/http/outbound/ReactiveHttpRequestExecutingMessageHandler.java [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+			return responseMono.map(response ->
+					new ResponseEntity<>(responseType != null
+							? response.body(BodyExtractors.toMono(responseType)).block()
+							: null,
+							response.headers().asHttpHeaders(),
+							response.statusCode()))
+					.map(this::getReply);
+		}
+		else {
+			responseMono.block();
```
Comment (garyrussell) [inline spring-integration-http/src/main/java/org/springframework/integration/http/outbound/ReactiveHttpRequestExecutingMessageHandler.java]: Why do we block here? If it's reactive with no reply should it just be send and ignore?

File: spring-integration-http/src/main/resources/org/springframework/integration/http/config/spring-integration-http-5.0.xsd [hunk 4]
```diff
@@ -424,7 +424,19 @@
...
-			<xsd:attributeGroup ref="asyncHttpOutboundCommonAttributes"/>
+			<xsd:attribute name="web-client" type="xsd:string">
+				<xsd:annotation>
+					<xsd:appinfo>
+						<tool:annotation kind="ref">
+							<tool:expected-type type="org.springframework.web.reactive.function.client.WebClient" />
+						</tool:annotation>
+					</xsd:appinfo>
+					<xsd:documentation>
+						The reference to org.springframework.web.reactive.function.client.WebClient
```
Comment (garyrussell) [inline spring-integration-http/src/main/resources/org/springframework/integration/http/config/spring-integration-http-5.0.xsd]: A reference to a ... bean which is used to send ...

**Review by garyrussell — COMMENTED** [review]
That's all; I can do the polishing if you can just answer my question.

File: spring-integration-http/src/main/java/org/springframework/integration/http/outbound/ReactiveHttpRequestExecutingMessageHandler.java (hunk 3 above)
Comment (artembilan) [inline spring-integration-http/src/main/java/org/springframework/integration/http/outbound/ReactiveHttpRequestExecutingMessageHandler.java]: H-m-m. Good catch, agree. Well, without blocking we don't have ability to handle error if that.
Although, I think we can bring here the same solution that we have for the `ListenableFuture` and its `onFailure()`. I mean that `sendErrorMessage()`.

Will fix soon!
Thank you!

**Comment by artembilan** [comment]
@garyrussell , 

pushed.

File: spring-integration-http/src/main/java/org/springframework/integration/http/outbound/HttpRequestExecutingMessageHandler.java [hunk 5]
```diff
@@ -148,7 +149,7 @@ public void setRequestFactory(ClientHttpRequestFactory requestFactory) {
 	}
 
 	@Override
-	protected Object exchange(URI uri, HttpMethod httpMethod, HttpEntity<?> httpRequest, Object expectedResponseType) {
+	protected Object exchange(URI uri, HttpMethod httpMethod, HttpEntity<?> httpRequest, Object expectedResponseType, Message<?> requestMessage) {
```
Comment (garyrussell) [inline spring-integration-http/src/main/java/org/springframework/integration/http/outbound/HttpRequestExecutingMessageHandler.java]: Line length violation.

File: spring-integration-http/src/main/resources/org/springframework/integration/http/config/spring-integration-http-5.0.xsd [hunk 6]
```diff
@@ -432,8 +432,8 @@
 						</tool:annotation>
 					</xsd:appinfo>
 					<xsd:documentation>
-						The reference to org.springframework.web.reactive.function.client.WebClient
-						bean to send the HTTP Request.
+						A reference to a org.springframework.web.reactive.function.client.WebClient bean
+						which is used to send to send the HTTP Requests reactive manner.
 					</xsd:documentation>
```
Comment (garyrussell) [inline spring-integration-http/src/main/resources/org/springframework/integration/http/config/spring-integration-http-5.0.xsd]: Other place too.

**Review by garyrussell — COMMENTED** [review]
That's all; will fix on merge.

File: spring-integration-http/src/main/java/org/springframework/integration/http/outbound/ReactiveHttpRequestExecutingMessageHandler.java [hunk 7]
```diff
@@ -0,0 +1,198 @@
...
+			responseMono
+					.doOnNext(response -> {
+						HttpStatus httpStatus = response.statusCode();
+						if (httpStatus.is4xxClientError() || httpStatus.is5xxServerError()) {
+							throw new WebClientException(
+									"ClientResponse has erroneous status code: " + httpStatus.value() +
+											" " + httpStatus.getReasonPhrase());
+						}
+					})
+					.subscribe(v -> {}, ex -> sendErrorMessage(requestMessage, ex));
```
Comment (artembilan) [inline spring-integration-http/src/main/java/org/springframework/integration/http/outbound/ReactiveHttpRequestExecutingMessageHandler.java]: Uh! There is a couple of Checkstyle violations, too:
```
:spring-integration-http:checkstyleMain[ant:checkstyle] [ERROR] /home/travis/build/spring-projects/spring-integration/spring-integration-http/src/main/java/org/springframework/integration/http/outbound/ReactiveHttpRequestExecutingMessageHandler.java:192:58: '{' is not followed by whitespace. [WhitespaceAround]

[ant:checkstyle] [ERROR] /home/travis/build/spring-projects/spring-integration/spring-integration-http/src/main/java/org/springframework/integration/http/outbound/ReactiveHttpRequestExecutingMessageHandler.java:192:58: '}' is not preceded with whitespace. [WhitespaceAround]
```

This must be like `v -> { }` 😄 

Please, fix this on merge as well, @garyrussell

**Comment by garyrussell** [comment]
Merged as 19c5402079e3f8e35bbe7db4ab1aa9b134db17a1 with minor polish and checkstyle fix.

**Pull request closed without merging**